lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs

```diff
@@ -1,19 +1,73 @@
 {-# LANGUAGE DataKinds #-}
 {-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
 
 module Cardano.Wallet.Jormungandr.Transaction
     ( newTransactionLayer
     ) where
 
 import Prelude
 
+import Cardano.Wallet.Jormungandr.Compatibility
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (AddressK), Key, Passphrase (..), XPrv, getKey )
+import Cardano.Wallet.Primitive.Types
+    ( Hash (..), Tx (..), TxId (..), TxOut (..), TxWitness (..) )
 import Cardano.Wallet.Transaction
-    ( TransactionLayer (..) )
+    ( ErrMkStdTx (..), TransactionLayer (..) )
+import Control.Monad
+    ( forM )
+import Data.ByteString
+    ( ByteString )
+import Data.Quantity
+    ( Quantity (..) )
 
+import qualified Cardano.Crypto.Wallet as CC
 
 -- | Construct a 'TransactionLayer' compatible with Shelley and 'Jörmungandr'
-newTransactionLayer :: TransactionLayer t
-newTransactionLayer = TransactionLayer
-    { mkStdTx = error "TODO: See http-bridge as starting point"
-    , estimateSize = error "TODO: See http-bridge as starting point"
+newTransactionLayer
+    :: forall n . Hash "Block0Hash"
+    -> TransactionLayer (Jormungandr n)
+newTransactionLayer block0Hash = TransactionLayer
+    { mkStdTx = \keyFrom inps outs -> do
+        let ins = (fmap fst inps)
+        let tx = Tx ins outs
+        let witData = witnessUtxoData block0Hash (txId @(Jormungandr n) tx)
+        txWitnesses <- forM inps $ \(_in, TxOut addr _c) -> mkWitness witData
+            <$> withEither (ErrKeyNotFoundForAddress addr) (keyFrom addr)
+        return (tx, txWitnesses)
+
+    -- NOTE: at this point 'Jörmungandr' node does not support fee calculation
+    , estimateSize = \_ -> Quantity 0
     }
+  where
+    withEither :: e -> Maybe a -> Either e a
+    withEither e = maybe (Left e) Right
+
+newtype WitnessData = WitnessData ByteString
+
+-- Rust implementation:
+--
+-- https://github.com/input-output-hk/rust-cardano/blob/13af7aa08cdac8efc9dc8b0f7ff9162d28b2d479/chain-impl-mockchain/src/transaction/witness.rs#L60-L61
+witnessUtxoData :: Hash "Block0Hash" -> Hash "Tx" -> WitnessData
```

Comment:
minor stuff, but maybe we would like to have all functions below in `where` of newTransactionLayer` ???

Reply:
I was thinking one _might_ want to export and test individually. And `witnessUtxoData` will be (one of several) "constructors" for `WitnessData`, so I think they should be on the same level as the `WitnessData` declaration.

Happy to surrender though. 

cc @KtorZ 

> Why the WitnessData indirection here? Why not have the hashes as arguments directly?